tests/component/test_time.py

```diff
@@ -0,0 +1,277 @@
+    random.shuffle(shuffled)
+    assert sorted(shuffled) == ordered
+
+
+@pytest.mark.parametrize("from_dt", [(JAN_01_2022_DATETIME), (JAN_01_2022_HIGHTIME)])
+def test___libtimestamp___convert_from_utc___succeeds(from_dt):
+    ts = LibTimestamp.from_datetime(from_dt)
+    assert ts == JAN_01_2022_LIB
+
+
+def test___libtimestamp___convert_to_utc___succeeds():
+    dt = JAN_01_2022_LIB.to_datetime(tzinfo=timezone.utc)
+    assert dt == JAN_01_2022_HIGHTIME
+
+
+@pytest.mark.parametrize(
+    "datetime_cls, tzinfo, expected_offset",
+    [
+        (std_datetime, timezone(timedelta(minutes=30)), -1800),
+        (std_datetime, timezone(timedelta(minutes=-30)), 1800),
+        (std_datetime, timezone(timedelta(hours=1)), -3600),
+        (std_datetime, timezone(timedelta(hours=-1)), 3600),
+        (ht_datetime, timezone(timedelta(minutes=30)), -1800),
+        (ht_datetime, timezone(timedelta(minutes=-30)), 1800),
+        (ht_datetime, timezone(timedelta(hours=1)), -3600),
+        (ht_datetime, timezone(timedelta(hours=-1)), 3600),
+    ],
+)
+def test___libtimestamp___convert_to_and_from_tz___succeeds(datetime_cls, tzinfo, expected_offset):
+    from_dt = datetime_cls(2002, 1, 1, tzinfo=tzinfo)
+    ts = LibTimestamp.from_datetime(from_dt)
+
+    assert ts.msb == JAN_01_2022_LIB.msb + expected_offset
+    assert ts.lsb == JAN_01_2022_LIB.lsb
+
+    # now convert back
+    to_dt = ts.to_datetime(tzinfo=tzinfo)
+    assert from_dt == to_dt
+
+
+@pytest.mark.parametrize(
+    "from_dt, microsecond, subseconds",
+    [
+        (JAN_01_2022_DATETIME, 0, 0),
+        (JAN_01_2022_DATETIME, 1, 0x10C6F7A0B5EE),
+        (JAN_01_2022_DATETIME, 250000, 0x4000000000000000),
+        (JAN_01_2022_DATETIME, 500000, 0x8000000000000000),
+        (JAN_01_2022_DATETIME, 750000, 0xC000000000000000),
+        (JAN_01_2022_DATETIME, 999999, 0xFFFFEF39085F4800),
+        (JAN_01_2022_HIGHTIME, 0, 0),
+        (JAN_01_2022_HIGHTIME, 1, 0x10C6F7A0B5EE),
+        (JAN_01_2022_HIGHTIME, 250000, 0x4000000000000000),
+        (JAN_01_2022_HIGHTIME, 500000, 0x8000000000000000),
+        (JAN_01_2022_HIGHTIME, 750000, 0xC000000000000000),
+        (JAN_01_2022_HIGHTIME, 999999, 0xFFFFEF39085F4800),
+    ],
+)
+def test___libtimestamp___convert_microseconds_to_and_from_subseconds___succeeds(
+    from_dt, microsecond, subseconds
+):
+    precise_dt = from_dt.replace(microsecond=microsecond)
+    ts = LibTimestamp.from_datetime(precise_dt)
+
+    # whole seconds shouldn't change
+    assert ts.msb == JAN_01_2022_LIB.msb
+    assert ts.lsb == subseconds
+
+    # now convert back
+    to_dt = ts.to_datetime(tzinfo=timezone.utc)
+
+    to_dt_microsecond = to_dt.microsecond
+    # comparison is tricky since imprecision in the conversion to NI-BTF are
+    # caught by the higher precision values in hightime, so we round here.
+    if to_dt.femtosecond > LibTimestamp.MAX_FS / 2:
+        to_dt_microsecond += 1
+
+    assert to_dt_microsecond == microsecond
```

Comment:
The test body should have 3 blocks, separated by blank lines: arrange, act, assert

If the middle "when" clause contains "convert_to_and_from", then that code belongs in the middle "act" block.

Please look over the regression test conventions and unit test conventions again.

If you feel the need to write comments with a running commentary, consider writing helper functions instead.

Reply:
> Please look over the regression test conventions and unit test conventions again.

Good recommendation! I think I have it better now, but I have low confidence it is correct. Thanks for the help!